### PR TITLE
CBG-3764 create multi-xattr subdoc APIs

### PR DESCRIPTION
**Pull request by torcolvin**
Update multi xattr subdoc APIs to support multiple xattrs. Where the APIs previously accepted two xattrs (user xattr and sync xattr), continue to allow two xattrs with CBS 7.6 with a CAS check in a loop. If the API previously did not accept two APIs, Couchbase Server 7.2 will return an invalid argument error if multiple xattrs are not specified.

Possible enhancements:

- enhance rosmar to be able to toggle on `BucketStoreFeatureMultiXattrSubdocOperations` to be able to test without running two versions of couchbase server
- Writing a check in all xattr APIs to make sure there are only 1 or 2 supported xattrs if using CBS < 7.6 which provides a bit easier message than one.
- Write a check in all xattr APIs to make sure that `xattrKeys` are never empty string. This was a late fix in integration tests from a somewhat opaque gocb error message. Alternatively, error in rosmar if any `xattrKeys` are empty.
- Move `ErrXattrNotFound` and `ErrXattrPartialFound` to sg-bucket and return in rosmar?

Things to look for in the PR:

- Did I remove any operations that should be retried to work a single time?
- Should I rename any functions to make it more clear which is the function that will do a cas based retry and which does the actual operation?
- Double checking any cases where I change logging for redaction OR error wrapping.

https://github.com/couchbaselabs/rosmar/pull/25
https://github.com/couchbase/sg-bucket/pull/112

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## Dependencies (if applicable)
- [x] Link upstream PRs
- [x] Update Go module dependencies when merged

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true,CBS-7.2.2` https://jenkins.sgwdev.com/job/SyncGateway-Integration/2366/
- [x] `GSI=true,xattrs=true,CBS-7.6.1-3122` https://jenkins.sgwdev.com/job/SyncGateway-Integration/2364/

### CODE REVIEW

File: base/collection_xattr.go
Comment (adamcfraser) [inline base/collection_xattr.go]: Do we care that the mutateIn functions that iterate over xattrKeys don't have similar defensive checks for empty keys?  (it's fine if the answer is 'no we don't' - just wanted to get your thoughts on it).

File: db/database.go
Comment (adamcfraser) [inline db/database.go]: A helper function to build the xattr keys wouldn't be a bad idea, since we're going to need to enhance it to add _vv in the near future.

File: base/collection_xattr.go
Comment (adamcfraser) [inline base/collection_xattr.go:196]: I think a comment might be merited here to explain that this is backward compatibility for "one system xattr and one user xattr" support.

For that matter, should we return error in this case if xattrKey2 is a system xattr (underscore prefix), to prevent potential misuse?

File: base/collection_xattr.go
Comment (adamcfraser) [inline base/collection_xattr.go]: It's not immediately obvious that "Couchbase Server < 7.6" corresponds to `c.IsSupported(sgbucket.BucketStoreFeatureMultiXattrSubdocOperations` - maybe just change this comment to `If BucketStoreFeatureMultiXattrSubdocOperations not supported...`

**Review by adamcfraser — CHANGES_REQUESTED** [review]
A few minor comments/questions but otherwise looks good to me.

File: base/collection_xattr.go
Comment (torcolvin) [reply base/collection_xattr.go]: I think I'm inclined to drop this defensive check, the error message is fine from gocb `invalid xattr fetch with no path`. I had added this during debug.

When I went to look at adding this elsewhere, it seemed more confusing than not to add. The enhancement I think would be worth doing is making sure that we can't pass an empty xattr to rosmar https://issues.couchbase.com/browse/CBG-3858

File: base/collection_xattr.go
Comment (torcolvin) [reply base/collection_xattr.go:196]: I was worried about ordering here for these arguments, but I'm open to the idea. The other idea would be to create a non sg-bucket function but inside Sync Gateway for getting a system xattr and a non system xattr, so the caller would be aware that they are taking a performance penalty.

The order of the arguments is trivially checkable, and to me it is an implementation detail that the xattrs we are adding will be system xattrs but the performance cost is present for any time there is more than one xattr.

**Review by adamcfraser — DISMISSED** [review]

**Comment by adamcfraser** [comment]
LGTM - ready to merge once dependencies are merged and go.mod updated.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**
